When you change the storage driver we ignore the storage-options

rhatdan opened the pull request:
The storage driver and the storage options in storage.conf should
match, but if you change the storage driver via the command line
then we need to nil out the default storage options from storage.conf.

If the user wants to change the storage driver and use storage options,
they need to specify them on the command line.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

edsantiago commented:
Check with @mheon; there was some discussion earlier today about using per-driver opts using sections, e.g. `[storage.options.vfs]`, `[storage.options.overlay]`, etc; but it needs coordination with containers/storage.

rhatdan commented:
I would rather not complicate those matters,  It would become a ADMIN nightmare, unless we support multiple drivers simultaneously.   I think this fixes the issue until we decide to support multi drivers.

mheon commented on cmd/podman/libpodruntime/runtime.go:
```diff
@@ -67,7 +67,10 @@ func getRuntime(ctx context.Context, c *cliconfig.PodmanCommand, renumber bool,
 		storageSet = true
 		storageOpts.GraphDriverName = c.GlobalFlags.StorageDriver
+		// Overriding the default storage driver caused GraphDriverOptions from storage.conf to be ignored
+		storageOpts.GraphDriverOptions = []string{}
```
Does this actually fix it? I think it's Mount options that are the problem, not graph driver options

rhatdan replied:
Ok that might be true.  But this is definitely a bug that should be fixed.

rhatdan replied:
After looking at the code.  mountopt is a storageopt, so this should fix the issue.

cevich commented:
@mheon @rhatdan With this applied, I'm still seeing [failures in F30 when running rootless](https://cirrus-ci.com/task/5609254338166784). The error message is the same as before this PR, bit I'm not sure if cause is the same:  `vfs driver does not support mount options`.

edsantiago commented:
If you like one line up from the error you'll see
```
[+0381s] Running: .../podman --storage-opt vfs.imagestore=/tmp/podman/imagecachedir [...] --storage-driver vfs
```
From a quick look this appears to be `test/utils/utils.go:77`, last touched by f610a485c by @baude. 

Does anyone know when exactly the vfs driver started griping about options? It sounds like that's new.

cevich commented:
Good catch Ed.

@baude perhaps when running rootless, those tests need to set `nocache=true`?  Though a simpler option may be to just not use the image cache when running rootless (I'm not sure what the de-optimization penalty is).

cevich commented:
Related, I also get [failures on normal, root-run F30 tests](https://cirrus-ci.com/task/5974003529809920).  Though perhaps worse, some give no error message relates to the storage driver :confused:

rhatdan commented:
@mheon @baude @cevich @edsantiago @TomSweeneyRedHat @giuseppe @vrothberg I don't see why this should not be merged?  
Can I get some LGTM. Or do we still need more changes?

cevich commented:
@rhatdan it seems to not  fix problem in Fedora 30...or we're interpreting the test results incorrectly.  I'll rebase #3078 against this PR and run again just to be double-sure...

edsantiago commented:
@rhatdan I think there's something else going on that I can't figure out. Basically, no matter what I do, I cannot reproduce the failure by using the `--storage-opt` command-line option:
```
# podman --storage-driver=vfs info
[all ok]
```
It's only when I hand-edit `/etc/containers/storage.conf`, change driver to "vfs", but leave `mountopt=` untouched, that's when I get the "vfs driver does not support mount options" error.

I would prefer to have a solid understanding of the problem before we merge this. (I'm also concerned about the approach you take here, because I'm not sure if it would clobber command-line `--storage-opts`)

edsantiago commented:
More weirdness:
```
# grep ^driver /etc/containers/storage.conf
driver = "vfs"                   ! hand-edited
# rm -rf /var/lib/containers;podman --storage-driver=overlay info
Error: could not get runtime: overlay: Unknown option vfs.mountopt
```
I obviously do not understand the `--storage-driver` option.

baude commented:
/hold

baude commented:
do not merge this yet

rhatdan commented:
@edsantiago If you edit storage.conf, it is expected you leave it in a consistent state.  So editing it to change driver to vfs but leaving the mount-options, then should cause an error.
which is why
```
# grep ^driver /etc/containers/storage.conf
driver = "vfs"                   ! hand-edited
# rm -rf /var/lib/containers;podman --storage-driver=overlay info
Error: could not get runtime: overlay: Unknown option vfs.mountopt
```
fails.

We have documented that overriding the storage-driver from the command line with --storage-driver, should cause the storage-options to be ignored as defined in storage.conf.

Looks like 
```
# rm -rf /tmp/ttt;podman --storage-driver=vfs --storage-opt=dsf --root=/tmp/ttt info|grep -1 Graph
    number: 0
  GraphDriverName: vfs
  GraphOptions:
  - dsf          <------ I would not expect to see options when vfs is used
  GraphRoot: /tmp/ttt
  GraphStatus: {}
  ImageStore:
```
Is because vfs driver only looks for specific mount options, and ignores others.
```
	for _, option := range options {
		if strings.HasPrefix(option, "vfs.imagestore=") {
			d.homes = append(d.homes, strings.Split(option[15:], ",")...)
			continue
		}
		if strings.HasPrefix(option, ".imagestore=") {
			d.homes = append(d.homes, strings.Split(option[12:], ",")...)
			continue
		}
		if strings.HasPrefix(option, "vfs.ostree_repo=") {
			if !ostree.OstreeSupport() {
				return nil, fmt.Errorf("vfs: ostree_repo specified but support for ostree is missing")
			}
			d.ostreeRepo = option[16:]
		}
		if strings.HasPrefix(option, ".ostree_repo=") {
			if !ostree.OstreeSupport() {
				return nil, fmt.Errorf("vfs: ostree_repo specified but support for ostree is missing")
			}
			d.ostreeRepo = option[13:]
		}
		if strings.HasPrefix(option, "vfs.mountopt=") {
			return nil, fmt.Errorf("vfs driver does not support mount options")
		}
	}
```

rhatdan commented:
So my patch should fix any use case other then when you make the storage.conf file inconsistent.

edsantiago commented:
@rhatdan thanks, I now understand the `--storage-opt` option better; but I still don't see what this patch accomplishes. `--storage-driver=vfs` does not seem to generate errors no matter what I try; only if I hand-edit storage.conf

baude commented:
im hoping to get root cause here, i have frozen the ci vm that does this testing so i can poke around when it finishes.

baude commented:
also, @rhatdan this appears to throw away --storage-opts provided by the user on the command line

rhatdan commented:
@baude works for me

```
 ./bin/podman --storage-opt=abcd --storage-driver=vfs --storage-opt=1234 --root=/tmp/ttt info 
...
  GraphDriverName: vfs
  GraphOptions:
  - abcd
  - "1234"
  GraphRoot: /tmp/ttt
  GraphStatus: {}
```

edsantiago commented:
@rhatdan now try that *without* your patch...

rhatdan commented:
@edsantiago Without this patch we get
```
  GraphDriverName: vfs
  GraphOptions:
  - overlay.mountopt=nodev,metacopy=on
```
Which does not cause an issue, but is still wrong according to the documention and could cause issues in the future.

rhatdan commented:
Without my patch specifying the graphdriver-options seems to override the storage.conf, which I am not sure is expected.

```
  GraphDriverName: vfs
  GraphOptions:
  - abcd
  - "1234"
```

rhatdan commented:
This behaviour matches the documentation.

```       --storage-opt=value

       Storage driver option, Default storage driver options are configured in
       /etc/containers/storage.conf ($HOME/.config/containers/storage.conf in
       rootless mode). The STORAGE_OPTS environment variable overrides the
       default. The --storage-opt specified options overrides all.
```
So I would say this is correct.

rhatdan commented:
@edsantiago Please test this out, to see if it fixes your --storage-driver=vfs --privileged problem?

edsantiago commented:
Addresses the problem as described, with reproducer, in #3263.

One behavior surprises me:
```
# ./bin/podman --conmon $CONMON_BINARY --root=/tmp/ttt --storage-driver=vfs --storage-opt=vfs.metacopy=on run --privileged alpine date
```
With the command-line `--storage-opt` override I was expecting to see the `does not support mount options` error; but the above works perfectly fine. LGTM either way, but I find this odd.

rhatdan commented:
@edsantiago we document that setting the --storage-driver overrules the storage-options set in the storage.conf file. So it ignores those settings, which is why it does not give any errors.

@edsantiago @mheon @baude @giuseppe @TomSweeneyRedHat I believe this is ready to merge.

mheon commented:
LGTM

edsantiago commented:
/lgtm

TomSweeneyRedHat commented:
All green, @mheon should we cancel @baude's hold on this?

mheon commented:
/hold cancel
